test/array.cpp

```diff
@@ -420,11 +426,290 @@ TEST(Array, ISSUE_951) {
     array b       = a.cols(0, 20).rows(10, 20);
 }
 
+TEST(Array, ISSUE_3534_SubArrayofModifiedSubArray) {
+    // This works
+    // array a = range(dim4(5,5));
+    // a = a.rows(0,3).copy();
+    // Following assignment failed silently without above copy()
+    // a(0,0) = 1234;
+
+    // Testing rows
+    {
+        array a = range(dim4(5, 5));
+        af_array before_handle = a.get();
+        const float *before_dev_pointer = a.device<float>();
+
+        a       = a.rows(1, 4);
+        af_array after_handle = a.get();
+        const float *after_dev_pointer = a.device<float>();
+
+        // We expect a copy to be created, so the handle and 
+        // array pointers should be different
+        ASSERT_NE(before_handle, after_handle);
+        ASSERT_NE(before_dev_pointer, after_dev_pointer);
+
+        a(1, 1) = -1234;
+
+        array b = range(dim4(4, 5)) + 1.0;
+        b(1, 1) = -1234;
+
+        ASSERT_ARRAYS_EQ(a, b);
+    }
+
+    // Testing columns
+    {
+        array a                         = range(dim4(5, 5));
+        af_array before_handle          = a.get();
+        const float *before_dev_pointer = a.device<float>();
+
+        a                              = a.cols(1, 4);
+        af_array after_handle          = a.get();
+        const float *after_dev_pointer = a.device<float>();
+
+        // We expect a copy to be created, so the handle and
+        // array pointers should be different
+        ASSERT_NE(before_handle, after_handle);
+        ASSERT_NE(before_dev_pointer, after_dev_pointer);
```

Comment:
There is a potential for these pointers to be the same because the device pointer is linear even though its calling a sub array. This is not the case for the rows based tests above because the subarray would not be linear.